codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -69,409 +71,1060 @@
+                int numBytesToBeConsumed = 0;
+                if (frame instanceof Http2DataFrame) {
+                    numBytesToBeConsumed = ((Http2DataFrame) frame).initialFlowControlledBytes();
+                    allocHandle.lastBytesRead(numBytesToBeConsumed);
+                } else {
+                    allocHandle.lastBytesRead(ARBITRARY_MESSAGE_SIZE);
+                }
+                allocHandle.incMessagesRead(1);
+                pipeline().fireChannelRead(frame);
+
+                if (numBytesToBeConsumed != 0) {
+                    try {
+                        onBytesConsumed(ctx, stream, numBytesToBeConsumed);
+                    } catch (Http2Exception e) {
+                        pipeline().fireExceptionCaught(e);
+                    }
+                }
+            }
+
+            @Override
+            public void write(Object msg, final ChannelPromise promise) {
+                if (!isActive()) {
+                    ReferenceCountUtil.release(msg);
+                    promise.setFailure(CLOSED_CHANNEL_EXCEPTION);
+                    return;
+                }
+
+                // After this point its not possible to cancel a write anymore.
+                if (!promise.setUncancellable()) {
+                    ReferenceCountUtil.release(msg);
+                    return;
+                }
+
+                try {
+                    if (msg instanceof Http2StreamFrame) {
+                        Http2StreamFrame frame = validateStreamFrame((Http2StreamFrame) msg).stream(stream());
+
+                        if (!firstFrameWritten && !isStreamIdValid(stream().id())) {
+                            if (!(frame instanceof Http2HeadersFrame)) {
+                                ReferenceCountUtil.release(frame);
+                                promise.setFailure(
+                                        new IllegalArgumentException("The first frame must be a headers frame. Was: "
+                                        + frame.name()));
+                                return;
+                            }
+                            firstFrameWritten = true;
+                            ChannelFuture future = write0(frame);
+                            if (future.isDone()) {
+                                firstWriteComplete(future, promise);
+                            } else {
+                                future.addListener(new ChannelFutureListener() {
+                                    @Override
+                                    public void operationComplete(ChannelFuture future) throws Exception {
+                                        firstWriteComplete(future, promise);
+                                    }
+                                });
+                            }
+                            return;
+                        }
+                    } else if (!(msg instanceof Http2GoAwayFrame)) {
```

Comment:
TODO: I don't think we should support GoAway on the child chan. It's trivial to write it to the parent() instead, and then there isn't the confusion of "how is it different writing it on the parent vs child". It's also strange that a single stream can write something that brings down the entire connection.